libraries/freertos_plus/standard/tls/src/iot_tls.c

```diff
@@ -389,6 +389,89 @@ static int prvPrivateKeySigningCallback( void * pvContext,
 
 /*-----------------------------------------------------------*/
 
+/**
+ * @brief Helper for reading the specified certificate object, if present,
+ * out of storage, into RAM, and then into an mbedTLS certificate context
+ * object.
+ *
+ * @param[in] pxTlsContext Caller TLS context.
+ * @param[in] pcLabelName PKCS #11 certificate object label.
+ * @param[in] xClass PKCS #11 certificate object class.
+ * @param[out] pxCertificateContext Certificate context.
+ *
+ * @return Zero on success.
+ */
+static int prvReadCertificateIntoContext( TLSContext_t * pxTlsContext,
+                                          const char * pcLabelName,
+                                          CK_OBJECT_CLASS xClass,
+                                          mbedtls_x509_crt * pxCertificateContext )
+{
+    BaseType_t xResult = CKR_OK;
+    CK_ATTRIBUTE xTemplate = { 0 };
+    CK_OBJECT_HANDLE xCertObj = 0;
+
+    /* Get the handle of the certificate. */
+    xResult = xFindObjectWithLabelAndClass( pxTlsContext->xP11Session,
+                                            pcLabelName,
+                                            xClass,
+                                            &xCertObj );
+
+    if( ( CKR_OK == xResult ) && ( xCertObj == CK_INVALID_HANDLE ) )
+    {
+        xResult = CKR_OBJECT_HANDLE_INVALID;
+    }
+
+    /* Query the certificate size. */
+    if( 0 == xResult )
+    {
+        xTemplate.type = CKA_VALUE;
+        xTemplate.ulValueLen = 0;
+        xTemplate.pValue = NULL;
+        xResult = ( BaseType_t ) pxTlsContext->pxP11FunctionList->C_GetAttributeValue( pxTlsContext->xP11Session,
+                                                                                       xCertObj,
+                                                                                       &xTemplate,
+                                                                                       1 );
+    }
+
+    /* Create a buffer for the certificate. */
+    if( 0 == xResult )
+    {
+        xTemplate.pValue = pvPortMalloc( xTemplate.ulValueLen ); /*lint !e9079 Allow casting void* to other types. */
```

Comment:
Should the buffer be cleared, or do we really trust C_GetAttributeValue?

Reply:
C_GetAttributeValue updates the template.ulValueLen on the second call to contain the _exact_ number of bytes returned by that call, so clearing this memory should not be necessary as it is this length that should be used in future.